src/libraries/Native/Unix/System.Globalization.Native/pal_icushim.c

```diff
@@ -353,23 +395,50 @@ int32_t GlobalizationNative_LoadICU()
             return FALSE;
         }
     }
-#endif
+#endif // TARGET_WINDOWS || TARGET_OSX
 
-    // Get pointers to all the ICU functions that are needed
-#define PER_FUNCTION_BLOCK(fn, lib) \
-    c_static_assert_msg((sizeof(#fn) + MaxICUVersionStringLength + 1) <= sizeof(symbolName), "The symbolName is too small for symbol " #fn); \
-    sprintf(symbolName, #fn "%s", symbolVersion); \
-    fn##_ptr = (__typeof(fn)*)dlsym(lib, symbolName); \
-    if (fn##_ptr == NULL) { fprintf(stderr, "Cannot get symbol %s from " #lib "\nError: %s\n", symbolName, dlerror()); abort(); }
+    FOR_ALL_ICU_FUNCTIONS
+    return TRUE;
+}
+
+void GlobalizationNative_InitICUFunctions(void* icuuc, void* icuin, const char* version, const char* suffix)
+{
+    assert(icuuc != NULL);
+    assert(icuin != NULL);
+    assert(version != NULL);
+    
+    libicuuc = icuuc;
+    libicui18n = icuin;
+    int major = -1;
+    int minor = -1;
+    int build = -1;
+
+    char symbolName[SYMBOL_NAME_SIZE];
+    char symbolVersion[MaxICUVersionStringWithSuffixLength + 1]="";
+    char symbolSuffix[SYMBOL_CUSTOM_SUFFIX_SIZE]="";
 
+    sscanf(version, "%d.%d.%d", &major, &minor, &build);
+
+    if (suffix != NULL)
+    {
+#if defined(TARGET_WINDOWS)
+    sprintf_s(symbolSuffix, SYMBOL_CUSTOM_SUFFIX_SIZE, "_%s", suffix);
+#else
+    sprintf(symbolSuffix, "_%s", suffix);
 #endif
+    }
 
-    FOR_ALL_ICU_FUNCTIONS
-#undef PER_FUNCTION_BLOCK
+    if(!FindSymbolVersion(major, minor, build, symbolName, symbolVersion, symbolSuffix))
+    {
+        fprintf(stderr, "Could not find symbol: %s from libicuuc\n", symbolName);
+        abort();
+    }
 
-    return TRUE;
+    FOR_ALL_ICU_FUNCTIONS
```

Comment:
One thought here is that I might suggest calling an ICU API in order to make sure that it is actually loaded correctly and callable. Otherwise you might later on fail once you call an API that indirectly depends on ICU.

Reply:
isn't calling `GetProcAddress` and `dlsym` on loaded binaries (which resolved their dependencies) enough?

Reply:
Unfortunately, it depends...

If the ICU data is packaged up using a .dat file (ex: icudt67l.dat, etc.), then the actual loading of the data file into memory will be done lazily (ie: on demand) later on when you call an API that uses/needs the data. In a low memory situation, this can fail due to out-of-memory (OOM).

On the other hand, if the ICU data is packaged up using a DLL, then the data file is forcefully loaded during the `LoadLibrary` call, so you don't need to worry about lazy-loading later on.
(If you run out of memory then the LoadLibrary call itself will fail).

In the app local case, I think the data file will likely always be in a DLL, so it perhaps isn't a concern. However, for the Windows OS version of ICU, a .dat file is used, so this can lead to OOM failures possibly happening later. 

FWIW, this issue actually did arise with ChakraCore, as the Spartan tabs (IIRC) were run inside memory-restricted environments.

I don't know if low-memory environments are common with .NET Core, but what we ended up doing in ChakraCore was to call `ulocdata_getCLDRVersion` to force the data file to be loaded during start up.
You can see the comments/code here:
https://github.com/microsoft/ChakraCore/blob/2f70be8cc93b2687750b2d244d5af3bc3edeffc8/lib/Runtime/Library/IntlEngineInterfaceExtensionObject.cpp#L610
(Initially this used the API `u_init`, but that was changed due to issues with NodeJS removing data).

Reply:
FWIW, I'm not trying to make this anymore difficult than it already is.. I just wanted to point out one of the issues that ChakraCore ran into with ICU. 🙃

Reply:
Interesting, will give it a thought. @tarekgh might also have comments in this. 

Regarding: 
>If the ICU data is packaged up using a .dat file (ex: icudt67l.dat, etc.), then the actual loading of the data file into memory will be done lazily

That would potentially break this workaround, no? https://github.com/dotnet/runtime/pull/35383/files#diff-d6e2b55e4fad16e09dee3702e746e59eR46

Reply:
Thanks safern. :)

> That would potentially break this workaround, no?
> https://github.com/dotnet/runtime/pull/35383/files#diff-d6e2b55e4fad16e09dee3702e746e59eR46
>> `// In Linux we need to load libicudata first because libicuuc and libicui18n depend on it. In order for the loader to find`

Ah, sorry for the confusion, I think I wasn't really clear above.

If you have a "libicudata.so" or "icudt[0-9]{2}.dll" file then the data is packaged up inside the shared library (.so/.dll), and thus you (likely) won't have any .dat file at all to worry about [*].

This would only really be an issue with the Windows OS version of ICU, which is loaded here:
https://github.com/dotnet/runtime/blob/1db45ed953745c0e270d09df99828cb58772f4ad/src/libraries/Native/Unix/System.Globalization.Native/pal_icushim.c#L33

[*] The reason I say "likely" above is that if somebody _really_ wanted to, they could modify their own app-local ICU, and make the data library (.so/.dll) a stub, and still use a .dat file. However, I don't know how much you really want to care about supporting hacked-up versions of app-local ICU...

Reply:
What happen when calling ICU API which ends up failing because of OOM and then calling it again (or calling other ICU API)? does ICU will retry to load the data or it just fails all the way? 
Also, does the whole data is just one thing? or can we have multiple data files? 

I am asking because we can avoid the OOM situation by just calling one of the APIi right after we load and that will ensure loading the data I guess. but if we have multiple data files, we can run into the situation later again. maybe this will be ok for us to throw exceptions during that time.

Reply:
> What happen when calling ICU API which ends up failing because of OOM and then calling it again (or calling other ICU API)? does ICU will retry to load the data or it just fails all the way?

I believe it depends on the API. Though, from testing, I found that for most APIs the ICU will try again to load the data again and if success, then it will continue fine.

However, there are some unfortunate ICU API (often older ones) that don't have an error code parameter at all, so if an OOM failure happens then you don't know about it.

For example, consider the API [`uloc_toUnicodeLocaleType`](https://unicode-org.github.io/icu-docs/apidoc/released/icu4c/uloc_8h.html#a0261e5fbb0ea96e16d111b681426f2a4), it has no way to report back an a detailed error status. So if any failure happens it returns `nullptr`, meaning that you don't know if the failure was OOM, or if maybe the input was garbage, etc.

So I think it sort of depends on if you want to force any OOM failures to happen up-front, or if you are okay with maybe some mysterious error happening later on.

> Also, does the whole data is just one thing? or can we have multiple data files?

There is only one single data file for the Window OS version, so if it loads successfully then you don't have to worry about other files failing later, etc.

> I am asking because we can avoid the OOM situation by just calling one of the API right after we load and that will ensure loading the data I guess. but if we have multiple data files, we can run into the situation later again. maybe this will be ok for us to throw exceptions during that time.

Good point/question. For a customized/hacked app-local copy of ICU, you could have multiple data files, so you might have to deal with OOM later on.
But again, I don't know if this is really something that you want to support or not, as I think it would be really esoteric and unlikely to happen in the wild at all.

Reply:
I would say for now we should ignore the multiple data files case and we can deal with that later if we run into it especially when having OOM situation. And we can try one of the ICU APIs right after loading to ensure forcing loading the data and we should be good to go.

Reply:
@jefgen any API that returns an error code that you would recommend to validate this?

Reply:
I'd probably suggest using this API here: [`ulocdata_getCLDRVersion`](https://unicode-org.github.io/icu-docs/apidoc/released/icu4c/ulocdata_8h.html#af50ca74fc2c0a249cd280f47c217dc88).

Though if you don't want to add usage of a new API, since it looks like this one isn't used by .NET Core yet, then you could use the API: [`u_init`](https://unicode-org.github.io/icu-docs/apidoc/released/icu4c/uclean_8h.html#ab6a4215b45b2162889aa1ff01e491d4c).

Both of these public/stable APIs will make ICU load data and they give you back an error code that you can check. If they fail, then you know you're in a bad state (ex: can't load the data file) and can bail out before you get too far along and/or you can throw an appropriate error. (If you want to be fine grained w.r.t. to the failure, then you can check the error code for 
`U_MEMORY_ALLOCATION_ERROR`, or `U_FILE_ACCESS_ERROR`, etc.)

The reason I'd probably suggest using `ulocdata_getCLDRVersion` instead of `u_init` is that:
- The `u_init` API actually depends on code-page/converter data being present in the ICU data file. If somebody builds a smaller ICU data file and excludes code-pages (and this is a real scenario, which is what NodeJS does), then the `u_init` function will fail. In this case, it doesn't mean that the data is bad or you hit OOM at all, which is why I suggest avoiding this API.
- The `ulocdata_getCLDRVersion` API only depends on a single [entry](https://github.com/unicode-org/icu/blob/63e480dedcbff7208f21a6e2f516d6fbe215bc83/icu4c/source/data/misc/supplementalData.txt#L2107) in the data which contains the CLDR version number, which isn't likely to ever be removed at all in smaller data files. So you don't need to worry if maybe the app-local data has been trimmed/reduced/etc.

Note that you can actually completely ignore the `versionArray` value that is returned back -- all you care about is if the `status` parameter/argument is a failure. :)

Reply:
Awesome. I guess we can add `ulocdata_getCLDRVersion` to the APIs we use, shouldn't be a big deal.

Reply:
Done.